subsys/net/buf.c

```diff
@@ -542,7 +577,7 @@ struct net_buf *net_buf_clone(struct net_buf *buf, s32_t timeout)
 	/* If the pool supports data referencing use that. Otherwise
 	 * we need to allocate new data and make a copy.
 	 */
-	if (pool->alloc->cb->ref) {
+	if (pool->alloc->cb->ref && !(buf->flags & NET_BUF_EXTERNAL_DATA)) {
```

Comment:
Just to bring your attention here: I made it so net_buf_clone will do a full-copy of the ext data. But maybe it's not the behaviour we want: shall we copy or point again at the external data from original net_buf?

Reply:
I think copying is the right thing to do, since there's no way to signal to the owner of the data that there are now multiple users of it (i.e. it might think it's free to reuse it once it's done with the original net_buf)

Reply:
> But maybe it's not the behaviour we want: shall we copy or point again at the external data from original net_buf?

I yet need to look at this in detail, but IMHO even this question should be enough to make a point: there can be *different* types of external data.

Unless there's a desire to support them all at one, I'd say that the usecase of "constant external data in ROM" is one of the most beneficial to tackles (offer the most saving). Of course, static constant data doesn't need copying.

(Of course, while important, that usecase doesn't have to be implemented *first*, we can start with with a more conservative implementation.)

Reply:
@jhedberg that was my assumption as well.

@pfalcon This new allocation function is meant to solve the case where user will do a blocking recv() or send(). These functions do not let the possibility to tell whether buffer is constant or not. (there are ways to discover if a pointer is in rom or ram space etc... but it's quite ugly and requires to store the right address space at built time in net stack's code that will handle the user buffers.).

Reply:
> These functions do not let the possibility to tell whether buffer is constant or not.

Well, my point: eventually we'll need to get to that.

> (there are ways to discover if a pointer is in rom or ram space etc... but it's quite ugly and requires to store the right address space at built time in net stack's code that will handle the user buffers.)

I don't think it's ugly nor that it should be "in net stack's code". It should be a generic kernel API, something like is_rom_ptr(p) (likely, is is_rom_region(buf, sz)). The alternative is letting users specify payload type explicitly, like lwIP does that: http://www.nongnu.org/lwip/2_0_x/group__pbuf.html#gab7e0e32fcc292c0d7107721766ed92fb  . IMHO, we should support both ;-).

Again, if you don't think that's the matter to discuss within scope of this PR, fine, but we'll need to get to that eventually.

Reply:
I wouldn't do this with something like `is_rom_ptr()`. Instead I'd just add another flag to net_buf to let the user say that the data is constant (after all, it might be in a fixed RAM location for whatever reason, but still never be modified). So we'd have some new `net_buf_set_fixed_data()` API that'd have as a pre-requisite that the new flag this PR adds is set. Anyway, agreed that this is not really relevant for this PR.

Reply:
@pfalcon, @jhedberg : sounds like a valid addition for the future. First, a generic way to know if a buffer is from rom space (k_is_rom_buffer for instance. Hiding the awfulness of checking space is quite relevant knowing that some hw have multiple space like that), and then a net_buf flag that net stack could set accordingly. I agree that net_buf should not deal with this complexity at all, after all: even in net stack it will be used only on specific use case.

Reply:
@tbursztyka:

@jhedberg:

> Anyway, agreed that this is not really relevant for this PR.

With that in mind, and with my "sockets" hat on, a socket user would just call send(sock, buf, sz, 0), so to reap full optimization benefits for that (e.g. for non-blocking send case), there would need to be a way to classify where the buf ptr points to.